core/naive-strategy.go

```diff
@@ -201,6 +203,126 @@ func (st *NaiveStrategy) UnrelayedPackets(ctx context.Context, src, dst *Provabl
+	logger := GetChannelPairLogger(src, dst)
+	var (
+		srcPackets                  PacketInfoList
+		dstPackets                  PacketInfoList
+		srcLatestHeight             ibcexported.Height
+		srcLatestTimestamp          uint64
+		srcLatestFinalizedHeight    ibcexported.Height
+		srcLatestFinalizedTimestamp uint64
+		dstLatestHeight             ibcexported.Height
+		dstLatestTimestamp          uint64
+		dstLatestFinalizedHeight    ibcexported.Height
+		dstLatestFinalizedTimestamp uint64
+	)
+
+	if 0 < len(rp.Src) {
+		if h, err := dst.LatestHeight(ctx); err != nil {
+			logger.Error("fail to get dst.LatestHeight", err)
+			return err
+		} else {
+			dstLatestHeight = h
+		}
+
+		if t, err := dst.Timestamp(ctx, dstLatestHeight); err != nil {
+			logger.Error("fail to get dst.Timestamp of latestHeight", err)
+			return err
+		} else {
+			dstLatestTimestamp = uint64(t.UnixNano())
+		}
+
+		dstLatestFinalizedHeight = sh.GetLatestFinalizedHeader(dst.ChainID()).GetHeight()
+		if t, err := dst.Timestamp(ctx, dstLatestFinalizedHeight); err != nil {
+			logger.Error("fail to get dst.Timestamp of latestFinalizedHeight", err)
+			return err
+		} else {
+			dstLatestFinalizedTimestamp = uint64(t.UnixNano())
+		}
+	}
+	if 0 < len(rp.Dst) {
+		if h, err := src.LatestHeight(ctx); err != nil {
+			logger.Error("fail to get src.LatestHeight", err)
+			return err
+		} else {
+			srcLatestHeight = h
+		}
+		if t, err := src.Timestamp(ctx, srcLatestHeight); err != nil {
+			logger.Error("fail to get src.Timestamp of  latestHeight", err)
+			return err
+		} else {
+			srcLatestTimestamp = uint64(t.UnixNano())
+		}
+
+		srcLatestFinalizedHeight = sh.GetLatestFinalizedHeader(src.ChainID()).GetHeight()
+		if t, err := src.Timestamp(ctx, srcLatestFinalizedHeight); err != nil {
+			logger.Error("fail to get src.Timestamp of  latestFinalizedHeight", err)
+			return err
+		} else {
+			srcLatestFinalizedTimestamp = uint64(t.UnixNano())
+		}
+	}
+
+	isTimeout := func(p *PacketInfo, height ibcexported.Height, timestamp uint64) bool {
+		return (!p.TimeoutHeight.IsZero() && p.TimeoutHeight.LTE(height)) ||
+			(p.TimeoutTimestamp != 0 && p.TimeoutTimestamp <= timestamp)
+	}
+
+	var srcTimeoutPackets, dstTimeoutPackets []*PacketInfo
+	for i, p := range rp.Src {
+		if isTimeout(p, dstLatestFinalizedHeight, dstLatestFinalizedTimestamp) {
+			p.TimedOut = true
+			if src.Path().GetOrder() == chantypes.ORDERED {
+				//  For ordered channel, a timeout notification will cause the channel to be closed.
+				//  Packets proceeding the timeout packet is relayed first
+				// so that they can be proceeded before the channel is closed.
+				// In ordered channels, only the first timed-out packet is selected because
+				// a timeout notification will close the channel. Subsequent packets cannot
+				// be processed once the channel is closed.
+				if i == 0 {
```

Comment:
@dai1975 Sorry for late question, but ...

Why do we need this limitation (`i == 0`)?
What happen if we remove this limitation?

Reply:
Before executing `TimeoutPacket` on ChainA, we need the following conditions.
- All preceding packets from ChainA have been `RecvPacket`'ed on ChainB.
- The executions of `RecvPacket` all have been finalized on ChainB.

For checking these conditions, I think it is insufficient to check `i == 0`, right?

Reply:
I guess, only if the first packet of packets returned from `dst.UnreceivedPackets` is timed out, we should execute `TimeoutPacket` for the packet on `src`.